### PR TITLE
[GC] Removed deprecated gcAddedOutboundReference and related logic

### PR DESCRIPTION
**Pull request by agarwal-navin**
The detection of outbound references was moved from DDS to container runtime by https://github.com/microsoft/FluidFramework/pull/18456. It deperecated `gcAddedOutboundReference` which is no longer needed.
This PR removes `gcAddedOutboundReference` and related logic. It also makes `gcAddedOutboundRoute` requried, which was added to the data store -> container runtime boundary.

[AB#8159](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8159)

### CODE REVIEW

File: packages/dds/shared-object-base/src/sharedObject.ts
Comment (markfields) [inline packages/dds/shared-object-base/src/sharedObject.ts:736]: So during Summarization/GC we would have had a ton of "addedOutboundReference" calls... I wonder if that was providing some coverage that we're now missing, so we're seeing gcUnknownOutboundReferences.  Just a shot in the dark.

File: packages/dds/shared-object-base/src/test/serializer.spec.ts
Comment (markfields) [inline packages/dds/shared-object-base/src/test/serializer.spec.ts:8]: I guess this callback didn't really have test coverage - oh well, it seemed to work :)

File: experimental/dds/tree/src/migration-shim/shimDeltaConnection.ts
Comment (markfields) [inline experimental/dds/tree/src/migration-shim/shimDeltaConnection.ts:129]: @tyler-cai-microsoft - Do you remember what exactly your concern was here?  https://github.com/microsoft/FluidFramework/pull/17709/files#diff-965d573e32dba6b55c5312a8ae2a6d8b4457227860fa892644c8cc402e246cb3R57

I don't fully grasp the data migration scenario, when this shim would kick in, etc.  With the new scheme where the ContainerRuntime peeks into each op payload when processing, and tells GC about new references at that point, do you have any more concerns around data migration?

File: packages/runtime/test-runtime-utils/package.json
Comment (markfields) [inline packages/runtime/test-runtime-utils/package.json:145]: I can't believe these mocks are `@alpha`.  We should open a bug to get them internal.

File: experimental/dds/tree/src/migration-shim/shimDeltaConnection.ts
Comment (tyler-cai-microsoft) [reply experimental/dds/tree/src/migration-shim/shimDeltaConnection.ts:129]: I don't think so

File: packages/runtime/test-runtime-utils/package.json
Comment (markfields) [reply packages/runtime/test-runtime-utils/package.json:145]: AB#8169

File: packages/runtime/datastore/src/dataStoreRuntime.ts
Comment (markfields) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:822]: So we are totally confident that the n/n-1 requirement for DataStoreRuntime/ContainerRuntime has been met at this point? Seems obvious but just want to call it out.  That we won't support RC5 DataStoreRuntime + internal.8.0 ContainerRuntime.

(is internal.8.0 the right "old" version here?)

Comment (markfields) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:822]: It's nice that we never made it required before, btw :P

File: packages/dds/shared-object-base/src/sharedObject.ts
Comment (agarwal-navin) [reply packages/dds/shared-object-base/src/sharedObject.ts:736]: Could be. But I would say that the system is more correct now than it was with this change.

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2962]: nit: consider pulling into separate PR

File: packages/runtime/datastore/src/dataStoreRuntime.ts
Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:822]: Looks like your change was released on rc1 which was on Jan 8. We are currently at rc5 and 5 months later. That is enough time to support compat at that layer :-).

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2962]: You are right. This is left over from the change that I started doing but then pivoted to this.

File: packages/runtime/container-runtime/src/channelCollection.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/channelCollection.ts:532]: This might also be a source of gcUnknownOutboundReferences? Probably not though since nothing changed here, I think.  But when I was working on auto-recovery I was telling GC about new outbound routes from <root> to the thing, but those routes weren't real and didn't show up in the GC data (if it was actually referenced, the real reference from <some actual node> to the thing was there but couldn't map to the fake added route).

In this case though, maybe we're ok - does ContainerRuntime's getGCData include routes from ContainerRuntime -> Aliased things?

File: packages/runtime/container-runtime/src/channelCollection.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/channelCollection.ts]: As discussed, remove this

File: packages/runtime/container-runtime/src/channelCollection.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/channelCollection.ts:814]: Update the interface too to make this required?

File: packages/runtime/container-runtime/src/channelCollection.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/channelCollection.ts:1186]: Is this for super old "root" concept, that's replaced by aliasing?  Why do we still need this code, like what's the scenario?

**Review by markfields — APPROVED** [review]

**Review by markfields — APPROVED** [review]

File: packages/runtime/container-runtime/src/channelCollection.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/channelCollection.ts:532]: > But when I was working on auto-recovery I was telling GC about new outbound routes from to the thing, but those routes weren't real and didn't show up in the GC data (if it was actually referenced, the real reference from to the thing was there but couldn't map to the fake added route).

Not sure I follow what you mean here.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/channelCollection.ts:532]: > In this case though, maybe we're ok - does ContainerRuntime's getGCData include routes from ContainerRuntime -> Aliased things?

I double checked the code. getGCData does include these routes.

File: packages/runtime/container-runtime/src/channelCollection.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/channelCollection.ts:1186]: I am removing this in a follow up PR.

File: packages/runtime/container-runtime/src/channelCollection.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/channelCollection.ts:533]: I will, in a follow up PR>

File: packages/runtime/container-runtime/src/channelCollection.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/channelCollection.ts:814]: I think we should remove this paramater. Data store runtime doesn't use it. It calls the `addedGCOutboundRoute` method of data store context. channel collection could also directly call `addedOutboundRoute` of the parent context.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/channelCollection.ts]: I removed it. Please take a look.

**Review by markfields — APPROVED** [review]

**Comment by agarwal-navin** [comment]
@microsoft/fluid-cr-tree Can someone take a look at this PR? It seems the PR is blocked on getting an approval?

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**
